agent/internal/fluent_intg_test.go

```diff
@@ -0,0 +1,257 @@
+			StdType:     stringToPointer("stdout"),
+		},
+		{
+			TrialID:     trialID,
+			AgentID:     &agentID,
+			ContainerID: stringToPointer("goodcontainer"),
+			RankID:      intToPointer(2),
+			Level:       stringToPointer("ERROR"),
+			Log:         stringToPointer("Workload completed: <RUN_STEP (100 Batches): (580,6289,4)> (duration 9:99:99)\n"), // nolint:lll
+			StdType:     stringToPointer("stdout"),
+		},
+		{
+			TrialID:     trialID,
+			AgentID:     &agentID,
+			ContainerID: stringToPointer("goodcontainer"),
+			RankID:      intToPointer(3),
+			Log:         stringToPointer("urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7f29a414dd30>: Failed to establish a new connection: [Errno 110] END\n"), // nolint:lll
+			StdType:     stringToPointer("stdout"),
+		},
+	}
+	actual := `[rank=4] 
+[rank=1] INFO: Workload completed: <RUN_STEP (100 Batches): (580,6289,4)> (duration 0:00:01.496132)
+[rank=2] ERROR: Workload completed: <RUN_STEP (100 Batches): (580,6289,4)> (duration 9:99:99)
+[rank=3] urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7f29a414dd30>: Failed to establish a new connection: [Errno 110] END` // nolint:lll
+	return expected, actual
+}
+
+func mockLogAcceptor(t *testing.T) (chan model.TrialLog, func()) {
+	e := echo.New()
+	baseListener, err := net.Listen("tcp", fmt.Sprintf(":%d", 8080))
+	assert.NilError(t, err, "error starting mock master listener")
+	e.Listener = baseListener
+	logBuffer := make(chan model.TrialLog, 100)
+	e.POST("/trial_logs", func(ctx echo.Context) error {
+		body, err := ioutil.ReadAll(ctx.Request().Body)
+		if err != nil {
+			return err
+		}
+
+		var logs []model.TrialLog
+		if err = json.Unmarshal(body, &logs); err != nil {
+			return err
+		}
+
+		for _, l := range logs {
+			end := strings.Contains(*l.Log, "END")
+			logBuffer <- l
+			if end {
+				close(logBuffer)
+				break
+			}
+		}
+		return nil
+	})
+	go func() {
+		if err := e.StartServer(e.Server); err != nil {
+			t.Logf("mock master exited: %s", err)
+		}
+	}()
+	return logBuffer, func() {
+		if err := e.Close(); err != nil {
+			t.Logf("error closing mock master echo server: %s", err)
+		}
+		if err := e.Listener.Close(); err != nil {
+			t.Logf("error closing mock master listener: %s", err)
+		}
+	}
+}
+
+func runContainerWithLogs(t *testing.T, fakeLogs string, trialID, fluentPort int) {
+	// WHEN we start a container that logs with fluentbit as its log driver.
+	docker, err := client.NewClientWithOpts(client.FromEnv, client.WithVersion("1.40"))
+	assert.NilError(t, err, "error connecting to Docker daemon")
+
+	logs, err := docker.ImagePull(context.Background(), "busybox", types.ImagePullOptions{})
+	assert.NilError(t, err, "failed to pull image")
+	_, err = ioutil.ReadAll(logs)
+	assert.NilError(t, err, "failed to drain pull logs")
+
+	// AND set some known ENV values.
```

Comment:
blocking: Hmm, seems like there should be more of these comments in this function after this one?

Reply:
ah yeah, these made more sense probably when it was all in line.